default/python/universe_generation/starsystems.py

```diff
@@ -84,13 +84,37 @@ def generate_systems(pos_list, gsd):
                               % (position[0], position[1]))
             continue
         sys_list.append(system)
-        for orbit in range(fo.sys_get_num_orbits(system)):
-            # check for each orbit if a planet shall be created by determining planet size
-            planet_size = planets.calc_planet_size(star_type, orbit, gsd.planet_density, gsd.shape)
-            if planet_size in planets.planet_sizes:
-                # ok, we want a planet, determine planet type and generate the planet
-                planet_type = planets.calc_planet_type(star_type, orbit, planet_size)
-                if fo.create_planet(planet_size, planet_type, system, orbit, "") == fo.invalid_object():
-                    # create planet failed, report an error and try to continue with next orbit
-                    util.report_error("Python generate_systems: create planet in system %d failed" % system)
+
+        orbits = range(fo.sys_get_num_orbits(system))
+
+        if not planets.can_have_planets(star_type, orbits, gsd.planet_density, gsd.shape):
+            continue
+
+        # Try to generate planets in each orbit.
+        # If after each orbit is tried once there are no planets then
+        # keep trying until a single planet is placed.
+        # Except for black hole systems, which can be empty.
+
+        at_least_one_planet = False
+        random.shuffle(orbits)
+        for orbit in orbits:
+            if planets.generate_a_planet(system, star_type, orbit, gsd.planet_density, gsd.shape):
+                at_least_one_planet = True
+
+        if star_type == fo.starType.blackHole:
+            continue
+
+        ii_safety = 1000
+        while not at_least_one_planet and ii_safety > 0:
+            ii_safety -= 1
+            for orbit in orbits:
+                if planets.generate_a_planet(system, star_type, orbit, gsd.planet_density, gsd.shape):
+                    at_least_one_planet = True
```

Comment:
Since it was a double loop, I used `itertools.product` with `else:`.  I hope that is sufficiently elegant.